eigen3/include/ignition/math/eigen3/Util.hh

```diff
@@ -0,0 +1,152 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+#ifndef IGNITION_MATH_EIGEN3_UTIL_HH_
+#define IGNITION_MATH_EIGEN3_UTIL_HH_
+
+#include <vector>
+
+#include <Eigen/Geometry>
+#include <Eigen/Eigenvalues>
+
+#include <ignition/math/AxisAlignedBox.hh>
+#include <ignition/math/Matrix3.hh>
+#include <ignition/math/OrientedBox.hh>
+#include <ignition/math/Pose3.hh>
+#include <ignition/math/Quaternion.hh>
+#include <ignition/math/Vector3.hh>
+#include <ignition/math/eigen3/Conversions.hh>
+
+namespace ignition
+{
+  namespace math
+  {
+    namespace eigen3
+    {
+      /// \brief Get covariance matrix from a set of 3d vertices
+      /// https://github.com/isl-org/Open3D/blob/76c2baf9debd460900f056a9b51e9a80de9c0e64/cpp/open3d/utility/Eigen.cpp#L305
+      /// \param[in] _mesh a mesh of 3d vertices
+      /// \return Covariance matrix
+      inline Eigen::Matrix3d covarianceMatrix(
+        const std::vector<math::Vector3d> &_mesh)
+      {
+        Eigen::Matrix<double, 9, 1> cumulants;
+        cumulants.setZero();
+        for (const auto &vertex : _mesh)
+        {
+          const Eigen::Vector3d &point = math::eigen3::convert(vertex);
+          cumulants(0) += point(0);
+          cumulants(1) += point(1);
+          cumulants(2) += point(2);
+          cumulants(3) += point(0) * point(0);
+          cumulants(4) += point(0) * point(1);
+          cumulants(5) += point(0) * point(2);
+          cumulants(6) += point(1) * point(1);
+          cumulants(7) += point(1) * point(2);
+          cumulants(8) += point(2) * point(2);
+        }
+
+        cumulants /= static_cast<double>(_mesh.size());
+
+        Eigen::Matrix3d covariance;
+        covariance(0, 0) = cumulants(3) - cumulants(0) * cumulants(0);
+        covariance(1, 1) = cumulants(6) - cumulants(1) * cumulants(1);
+        covariance(2, 2) = cumulants(8) - cumulants(2) * cumulants(2);
+        covariance(0, 1) = cumulants(4) - cumulants(0) * cumulants(1);
+        covariance(1, 0) = covariance(0, 1);
+        covariance(0, 2) = cumulants(5) - cumulants(0) * cumulants(2);
+        covariance(2, 0) = covariance(0, 2);
+        covariance(1, 2) = cumulants(7) - cumulants(1) * cumulants(2);
+        covariance(2, 1) = covariance(1, 2);
+        return covariance;
+      }
+
+      /// \brief Get the oriented 3d bounding box of a mesh points using PCA
+      /// http://codextechnicanum.blogspot.com/2015/04/find-minimum-oriented-bounding-box-of.html
+      /// \param[in] _mesh a mesh of 3d vertices
+      /// \return Oriented 3D box
+      inline ignition::math::OrientedBoxd meshToOrientedBox(
+        const std::vector<math::Vector3d> &_mesh)
+      {
+        math::Vector3d mean;
+        for (const auto &point : _mesh)
+          mean += point;
+        mean /= static_cast<double>(_mesh.size());
```

Comment:
should we check the vector size and exit early if the vector is empty to avoid dividing by zero?